## Reading the xlsx file

In [1]:
import pandas as pd

In [2]:
eda_df=pd.read_excel("./EDA_Dataset.xlsx")
eda_df

ID  Year_Birth   Education Marital_Status   Income  Kidhome  \
0      5524        1957  Graduation         Single  58138.0        0   
1      2174        1954  Graduation         Single  46344.0        1   
2      4141        1965  Graduation       Together  71613.0        0   
3      6182        1984  Graduation       Together  26646.0        1   
4      5324        1981         PhD        Married  58293.0        1   
...     ...         ...         ...            ...      ...      ...   
2235  10870        1967  Graduation        Married  61223.0        0   
2236   4001        1946         PhD       Together  64014.0        2   
2237   7270        1981  Graduation       Divorced  56981.0        0   
2238   8235        1956      Master       Together  69245.0        0   
2239   9405        1954         PhD        Married  52869.0        1   

      Teenhome Dt_Customer  Recency  MntWines  ...  NumWebVisitsMonth  \
0            0  2012-09-04       58       635  ...                  7   
1            1  2014-03-08       38        11  ...                  5   
2            0  2013-08-21       26       426  ...                  4   
3            0  2014-02-10       26        11  ...                  6   
4            0  2014-01-19       94       173  ...                  5   
...        ...         ...      ...       ...  ...                ...   
2235         1  2013-06-13       46       709  ...                  5   
2236         1  2014-06-10       56       406  ...                  7   
2237         0  2014-01-25       91       908  ...                  6   
2238         1  2014-01-24        8       428  ...                  3   
2239         1  2012-10-15       40        84  ...                  7   

      AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  \
0                0             0             0             0             0   
1                0             0             0             0             0   
2                0             0             0             0             0   
3                0             0             0             0             0   
4                0             0             0             0             0   
...            ...           ...           ...           ...           ...   
2235             0             0             0             0             0   
2236             0             0             0             1             0   
2237             0             1             0             0             0   
2238             0             0             0             0             0   
2239             0             0             0             0             0   

      Complain  Z_CostContact  Z_Revenue  Response  
0            0              3         11         1  
1            0              3         11         0  
2            0              3         11         0  
3            0              3         11         0  
4            0              3         11         0  
...        ...            ...        ...       ...  
2235         0              3         11         0  
2236         0              3         11         0  
2237         0              3         11         0  
2238         0              3         11         0  
2239         0              3         11         1  

[2240 rows x 29 columns]

# Creating a duplicate copy of dataset to preserve the original dataset

In [3]:
dupl_df=eda_df.copy()
dupl_df

ID  Year_Birth   Education Marital_Status   Income  Kidhome  \
0      5524        1957  Graduation         Single  58138.0        0   
1      2174        1954  Graduation         Single  46344.0        1   
2      4141        1965  Graduation       Together  71613.0        0   
3      6182        1984  Graduation       Together  26646.0        1   
4      5324        1981         PhD        Married  58293.0        1   
...     ...         ...         ...            ...      ...      ...   
2235  10870        1967  Graduation        Married  61223.0        0   
2236   4001        1946         PhD       Together  64014.0        2   
2237   7270        1981  Graduation       Divorced  56981.0        0   
2238   8235        1956      Master       Together  69245.0        0   
2239   9405        1954         PhD        Married  52869.0        1   

      Teenhome Dt_Customer  Recency  MntWines  ...  NumWebVisitsMonth  \
0            0  2012-09-04       58       635  ...                  7   
1            1  2014-03-08       38        11  ...                  5   
2            0  2013-08-21       26       426  ...                  4   
3            0  2014-02-10       26        11  ...                  6   
4            0  2014-01-19       94       173  ...                  5   
...        ...         ...      ...       ...  ...                ...   
2235         1  2013-06-13       46       709  ...                  5   
2236         1  2014-06-10       56       406  ...                  7   
2237         0  2014-01-25       91       908  ...                  6   
2238         1  2014-01-24        8       428  ...                  3   
2239         1  2012-10-15       40        84  ...                  7   

      AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  \
0                0             0             0             0             0   
1                0             0             0             0             0   
2                0             0             0             0             0   
3                0             0             0             0             0   
4                0             0             0             0             0   
...            ...           ...           ...           ...           ...   
2235             0             0             0             0             0   
2236             0             0             0             1             0   
2237             0             1             0             0             0   
2238             0             0             0             0             0   
2239             0             0             0             0             0   

      Complain  Z_CostContact  Z_Revenue  Response  
0            0              3         11         1  
1            0              3         11         0  
2            0              3         11         0  
3            0              3         11         0  
4            0              3         11         0  
...        ...            ...        ...       ...  
2235         0              3         11         0  
2236         0              3         11         0  
2237         0              3         11         0  
2238         0              3         11         0  
2239         0              3         11         1  

[2240 rows x 29 columns]

## Size of data

In [4]:
dupl_df.shape

(2240, 29)

## Disk space

In [5]:
dupl_df.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2240 entries, 0 to 2239
Data columns (total 29 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   ID                   2240 non-null   int64         
 1   Year_Birth           2240 non-null   int64         
 2   Education            2240 non-null   object        
 3   Marital_Status       2240 non-null   object        
 4   Income               2216 non-null   float64       
 5   Kidhome              2240 non-null   int64         
 6   Teenhome             2240 non-null   int64         
 7   Dt_Customer          2240 non-null   datetime64[ns]
 8   Recency              2240 non-null   int64         
 9   MntWines             2240 non-null   int64         
 10  MntFruits            2240 non-null   int64         
 11  MntMeatProducts      2240 non-null   int64         
 12  MntFishProducts      2240 non-null   int64         
 13  MntSweetProducts     2240 non-nul

## Datatype

In [6]:
dupl_df.dtypes

ID                              int64
Year_Birth                      int64
Education                      object
Marital_Status                 object
Income                        float64
Kidhome                         int64
Teenhome                        int64
Dt_Customer            datetime64[ns]
Recency                         int64
MntWines                        int64
MntFruits                       int64
MntMeatProducts                 int64
MntFishProducts                 int64
MntSweetProducts                int64
MntGoldProds                    int64
NumDealsPurchases               int64
NumWebPurchases                 int64
NumCatalogPurchases             int64
NumStorePurchases               int64
NumWebVisitsMonth               int64
AcceptedCmp3                    int64
AcceptedCmp4                    int64
AcceptedCmp5                    int64
AcceptedCmp1                    int64
AcceptedCmp2                    int64
Complain                        int64
Z_CostContac

### Data is quite big having 29 columns and 2240 rows with datatypes:datetime64[ns](1), float64(1), int64(25), object(2) and consuming 507.6+kb all columns are not null

## Columns

In [7]:
dupl_df.columns
dupl_df.nunique()


ID                     2240
Year_Birth               59
Education                 5
Marital_Status            8
Income                 1974
Kidhome                   3
Teenhome                  3
Dt_Customer             663
Recency                 100
MntWines                776
MntFruits               158
MntMeatProducts         558
MntFishProducts         182
MntSweetProducts        177
MntGoldProds            213
NumDealsPurchases        15
NumWebPurchases          15
NumCatalogPurchases      14
NumStorePurchases        14
NumWebVisitsMonth        16
AcceptedCmp3              2
AcceptedCmp4              2
AcceptedCmp5              2
AcceptedCmp1              2
AcceptedCmp2              2
Complain                  2
Z_CostContact             1
Z_Revenue                 1
Response                  2
dtype: int64

In [8]:
dupl_df

ID  Year_Birth   Education Marital_Status   Income  Kidhome  \
0      5524        1957  Graduation         Single  58138.0        0   
1      2174        1954  Graduation         Single  46344.0        1   
2      4141        1965  Graduation       Together  71613.0        0   
3      6182        1984  Graduation       Together  26646.0        1   
4      5324        1981         PhD        Married  58293.0        1   
...     ...         ...         ...            ...      ...      ...   
2235  10870        1967  Graduation        Married  61223.0        0   
2236   4001        1946         PhD       Together  64014.0        2   
2237   7270        1981  Graduation       Divorced  56981.0        0   
2238   8235        1956      Master       Together  69245.0        0   
2239   9405        1954         PhD        Married  52869.0        1   

      Teenhome Dt_Customer  Recency  MntWines  ...  NumWebVisitsMonth  \
0            0  2012-09-04       58       635  ...                  7   
1            1  2014-03-08       38        11  ...                  5   
2            0  2013-08-21       26       426  ...                  4   
3            0  2014-02-10       26        11  ...                  6   
4            0  2014-01-19       94       173  ...                  5   
...        ...         ...      ...       ...  ...                ...   
2235         1  2013-06-13       46       709  ...                  5   
2236         1  2014-06-10       56       406  ...                  7   
2237         0  2014-01-25       91       908  ...                  6   
2238         1  2014-01-24        8       428  ...                  3   
2239         1  2012-10-15       40        84  ...                  7   

      AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  \
0                0             0             0             0             0   
1                0             0             0             0             0   
2                0             0             0             0             0   
3                0             0             0             0             0   
4                0             0             0             0             0   
...            ...           ...           ...           ...           ...   
2235             0             0             0             0             0   
2236             0             0             0             1             0   
2237             0             1             0             0             0   
2238             0             0             0             0             0   
2239             0             0             0             0             0   

      Complain  Z_CostContact  Z_Revenue  Response  
0            0              3         11         1  
1            0              3         11         0  
2            0              3         11         0  
3            0              3         11         0  
4            0              3         11         0  
...        ...            ...        ...       ...  
2235         0              3         11         0  
2236         0              3         11         0  
2237         0              3         11         0  
2238         0              3         11         0  
2239         0              3         11         1  

[2240 rows x 29 columns]

## Conclusion by reading the dataset and column names
We can assume this dataset as a complete customer expolaratory record of a supermarket having unique customer ids, birth year,education,income etc.


In [9]:
dupl_df["Complain"].unique()


array([0, 1], dtype=int64)

In [10]:
#Checking all the unique values of columns
pd.Series({c: dupl_df[c].unique() for c in dupl_df})

ID                     [5524, 2174, 4141, 6182, 5324, 7446, 965, 6177...
Year_Birth             [1957, 1954, 1965, 1984, 1981, 1967, 1971, 198...
Education                     [Graduation, PhD, Master, Basic, 2n Cycle]
Marital_Status         [Single, Together, Married, Divorced, Widow, A...
Income                 [58138.0, 46344.0, 71613.0, 26646.0, 58293.0, ...
Kidhome                                                        [0, 1, 2]
Teenhome                                                       [0, 1, 2]
Dt_Customer            [2012-09-04 00:00:00, 2014-03-08 00:00:00, 201...
Recency                [58, 38, 26, 94, 16, 34, 32, 19, 68, 11, 59, 8...
MntWines               [635, 11, 426, 173, 520, 235, 76, 14, 28, 5, 6...
MntFruits              [88, 1, 49, 4, 43, 42, 65, 10, 0, 5, 16, 61, 2...
MntMeatProducts        [546, 6, 127, 20, 118, 98, 164, 56, 24, 11, 48...
MntFishProducts        [172, 2, 111, 10, 46, 0, 50, 3, 1, 11, 225, 6,...
MntSweetProducts       [88, 1, 21, 3, 27, 42, 49, 2

In [11]:
dupl_df[dupl_df.apply(pd.Series.nunique, axis=1) == 1]
dupl_df["Complain"].equals(dupl_df["AcceptedCmp3"])
#AcceptedCmp1,2,3,4,5 not related to complain


False

## Checking the conclusion about the cvolumns AcceptedCmp1,2,3,4,5 and response

## Assumption- If any of AcceptedCmp1,2,3,4,5 is true then Response is true


In [12]:
any_accepted = (dupl_df['AcceptedCmp1'] == 1) | (dupl_df['AcceptedCmp2'] == 1) | (dupl_df['AcceptedCmp3'] == 1) | (dupl_df['AcceptedCmp4'] == 1) | (dupl_df['AcceptedCmp5'] == 1)
response_accp = (dupl_df['Response'] == 1)
result = any_accepted & response_accp
dupl_df[result]


ID  Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
15    2114        1946         PhD         Single  82800.0        0         0   
39    2968        1943         PhD       Divorced  48948.0        0         0   
53    2225        1977  Graduation       Divorced  82582.0        0         0   
55    6260        1955      Master       Together  82384.0        0         0   
60    6853        1982      Master         Single  75777.0        0         0   
...    ...         ...         ...            ...      ...      ...       ...   
2175  1772        1975         PhD        Married  79174.0        0         0   
2193  8722        1957    2n Cycle        Married  82347.0        0         0   
2194  7118        1957  Graduation        Married  73803.0        0         1   
2198  2632        1954  Graduation        Married  50501.0        1         1   
2221  7366        1982      Master         Single  75777.0        0         0   

     Dt_Customer  Recency  MntWines  ...  NumWebVisitsMonth  AcceptedCmp3  \
15    2012-11-24       23      1006  ...                  3             0   
39    2013-02-01       53       437  ...                  6             1   
53    2014-06-07       54       510  ...                  1             1   
55    2012-11-19       55       984  ...                  1             0   
60    2013-07-04       12       712  ...                  1             0   
...          ...      ...       ...  ...                ...           ...   
2175  2013-01-11        2      1074  ...                  2             0   
2193  2012-11-06       38       556  ...                  3             1   
2194  2012-08-01       61       833  ...                  6             1   
2198  2013-06-18       10       297  ...                  6             1   
2221  2013-07-04       12       712  ...                  1             0   

      AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  Complain  \
15               0             1             1             0         0   
39               0             0             0             0         0   
53               0             0             1             0         0   
55               0             1             0             0         0   
60               1             1             0             0         0   
...            ...           ...           ...           ...       ...   
2175             0             1             1             0         0   
2193             0             0             1             0         0   
2194             0             0             0             0         0   
2198             0             0             0             0         0   
2221             1             1             0             0         0   

      Z_CostContact  Z_Revenue  Response  
15                3         11         1  
39                3         11         1  
53                3         11         1  
55                3         11         1  
60                3         11         1  
...             ...        ...       ...  
2175              3         11         1  
2193              3         11         1  
2194              3         11         1  
2198              3         11         1  
2221              3         11         1  

[188 rows x 29 columns]

In [13]:
any_accepted = (dupl_df['AcceptedCmp1'] == 0) | (dupl_df['AcceptedCmp2'] == 0) | (dupl_df['AcceptedCmp3'] == 0) | (dupl_df['AcceptedCmp4'] == 0) | (dupl_df['AcceptedCmp5'] == 0)
response_accp = (dupl_df['Response'] == 1)
result = any_accepted & response_accp
dupl_df[result]


ID  Year_Birth   Education Marital_Status   Income  Kidhome  \
0      5524        1957  Graduation         Single  58138.0        0   
8      4855        1974         PhD       Together  30351.0        1   
15     2114        1946         PhD         Single  82800.0        0   
33     7373        1952         PhD       Divorced  46610.0        0   
39     2968        1943         PhD       Divorced  48948.0        0   
...     ...         ...         ...            ...      ...      ...   
2194   7118        1957  Graduation        Married  73803.0        0   
2198   2632        1954  Graduation        Married  50501.0        1   
2202  11133        1973         PhD           YOLO  48432.0        0   
2221   7366        1982      Master         Single  75777.0        0   
2239   9405        1954         PhD        Married  52869.0        1   

      Teenhome Dt_Customer  Recency  MntWines  ...  NumWebVisitsMonth  \
0            0  2012-09-04       58       635  ...                  7   
8            0  2013-06-06       19        14  ...                  9   
15           0  2012-11-24       23      1006  ...                  3   
33           2  2012-10-29        8        96  ...                  6   
39           0  2013-02-01       53       437  ...                  6   
...        ...         ...      ...       ...  ...                ...   
2194         1  2012-08-01       61       833  ...                  6   
2198         1  2013-06-18       10       297  ...                  6   
2202         1  2012-10-18        3       322  ...                  8   
2221         0  2013-07-04       12       712  ...                  1   
2239         1  2012-10-15       40        84  ...                  7   

      AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  \
0                0             0             0             0             0   
8                0             0             0             0             0   
15               0             0             1             1             0   
33               0             0             0             0             0   
39               1             0             0             0             0   
...            ...           ...           ...           ...           ...   
2194             1             0             0             0             0   
2198             1             0             0             0             0   
2202             0             0             0             0             0   
2221             0             1             1             0             0   
2239             0             0             0             0             0   

      Complain  Z_CostContact  Z_Revenue  Response  
0            0              3         11         1  
8            0              3         11         1  
15           0              3         11         1  
33           0              3         11         1  
39           0              3         11         1  
...        ...            ...        ...       ...  
2194         0              3         11         1  
2198         0              3         11         1  
2202         0              3         11         1  
2221         0              3         11         1  
2239         0              3         11         1  

[334 rows x 29 columns]

## Our Above assumption is wrong as there are 334 rows having AcceptedCmp1,2,3,4,5 all false but Response as True

## Now we can assume Response as customer accepted the last offer or not

## Data Dictionary
1. ID                     2240 - Unique customer ids
2. Year_Birth               59 - Birthdate of customers
3. Education                 5 - Education qualification of customers
4. Marital_Status            8 - Married/single/divorced/together
5. Income                 1974 - Income of customer
6. Kidhome                   3 - no. of Kids in a house(below 13yrs)
7. Teenhome                  3 - no. of Teenagers(13 to 19 yrs)
8. Dt_Customer             663 - Date of first purchase/membership enrollment
9. Recency                 100 - No. of visits in last 1 year
10. MntWines                776 - Amnt spent on wines
11. MntFruits               158 -Amnt spent on fruits
12. MntMeatProducts         558- Amnt spent on Meat
13. MntFishProducts         182- Amnt spent on fish prod
14. MntSweetProducts        177- Amnt spent on sweet products
15. MntGoldProds            213- Amnt spent on gold products
16. NumDealsPurchases        15 - Number of deals purchased
17. NumWebPurchases          15 - No. of purchases from website of supermarket
18. NumCatalogPurchases      14- No. of purchases from website of catalog
19. NumStorePurchases        14-No. of purchases by offline visit of supermarket
20. NumWebVisitsMonth        16- Avg. No. of visits in a month
21. AcceptedCmp3              2 - Accepted offer in 3rd attempt or not
22. AcceptedCmp4              2 - Accepted offer in 4th attempt or not
23. AcceptedCmp5              2- Accepted offer in 5th attempt or not
24. AcceptedCmp1              2- Accepted offer in single attempt or not
25. AcceptedCmp2              2 - Accepted offer in 2nd attempt or not
26. Complain                  2 - 1 for complain raised and  for not raised
27. Z_CostContact             1 - Unknown- all duplicate
28. Z_Revenue                 1 - Unknown- all duplicate
29. Response                  2 - Aceepted the last offer or not


## Checking for missing values in dataset

In [14]:
dupl_df.isna()

ID  Year_Birth  Education  Marital_Status  Income  Kidhome  Teenhome  \
0     False       False      False           False   False    False     False   
1     False       False      False           False   False    False     False   
2     False       False      False           False   False    False     False   
3     False       False      False           False   False    False     False   
4     False       False      False           False   False    False     False   
...     ...         ...        ...             ...     ...      ...       ...   
2235  False       False      False           False   False    False     False   
2236  False       False      False           False   False    False     False   
2237  False       False      False           False   False    False     False   
2238  False       False      False           False   False    False     False   
2239  False       False      False           False   False    False     False   

      Dt_Customer  Recency  MntWines  ...  NumWebVisitsMonth  AcceptedCmp3  \
0           False    False     False  ...              False         False   
1           False    False     False  ...              False         False   
2           False    False     False  ...              False         False   
3           False    False     False  ...              False         False   
4           False    False     False  ...              False         False   
...           ...      ...       ...  ...                ...           ...   
2235        False    False     False  ...              False         False   
2236        False    False     False  ...              False         False   
2237        False    False     False  ...              False         False   
2238        False    False     False  ...              False         False   
2239        False    False     False  ...              False         False   

      AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  Complain  \
0            False         False         False         False     False   
1            False         False         False         False     False   
2            False         False         False         False     False   
3            False         False         False         False     False   
4            False         False         False         False     False   
...            ...           ...           ...           ...       ...   
2235         False         False         False         False     False   
2236         False         False         False         False     False   
2237         False         False         False         False     False   
2238         False         False         False         False     False   
2239         False         False         False         False     False   

      Z_CostContact  Z_Revenue  Response  
0             False      False     False  
1             False      False     False  
2             False      False     False  
3             False      False     False  
4             False      False     False  
...             ...        ...       ...  
2235          False      False     False  
2236          False      False     False  
2237          False      False     False  
2238          False      False     False  
2239          False      False     False  

[2240 rows x 29 columns]

## Percentage of missing values in dataset in each column

In [15]:
dupl_df.isna().sum()

ID                      0
Year_Birth              0
Education               0
Marital_Status          0
Income                 24
Kidhome                 0
Teenhome                0
Dt_Customer             0
Recency                 0
MntWines                0
MntFruits               0
MntMeatProducts         0
MntFishProducts         0
MntSweetProducts        0
MntGoldProds            0
NumDealsPurchases       0
NumWebPurchases         0
NumCatalogPurchases     0
NumStorePurchases       0
NumWebVisitsMonth       0
AcceptedCmp3            0
AcceptedCmp4            0
AcceptedCmp5            0
AcceptedCmp1            0
AcceptedCmp2            0
Complain                0
Z_CostContact           0
Z_Revenue               0
Response                0
dtype: int64

In [ ]:
(dupl_df.isna().sum()/len(dupl_df))*100

# Handling missing values and droping duplicate values

In [ ]:
missing_values = dupl_df[eda_df.isnull().any(axis=1)]

# Display the rows containing missing values
print(missing_values)

# Filling NaN values with mean income

In [ ]:
income_mean = dupl_df['Income'].mean()
dupl_df['Income'].fillna(income_mean, inplace=True)
dupl_df


In [ ]:
dupl_df.isna().sum()


In [ ]:
columns_to_drop = ['Z_Revenue', 'Z_CostContact']
dupl_df.drop(columns_to_drop, axis=1, inplace=True)

In [ ]:
dupl_df

## statistical properties of columns having numerical values as well as at least 5 unique column values.

In [ ]:
# Duplicating the dataset and then droping the Ids columns as ids has no role in statistical properties
dup_sdf=dupl_df.copy()
dup_sdf.drop('ID', axis=1, inplace=True)


In [ ]:
dup_sdf

In [ ]:
filtered_columns = dup_sdf.columns[dup_sdf.nunique() >= 5]

# Calculate statistical properties for the filtered columns
statistics = dup_sdf[filtered_columns].describe()
statistics

## Correlation

In [ ]:
##Slicing Id as it doesnt effect correlation and columns having strings/object
dupl_df.iloc[:, ~dupl_df.columns.isin(dupl_df.columns[[0, 2, 3]])]

In [ ]:
dupl_df.iloc[:, ~dupl_df.columns.isin(dupl_df.columns[[0, 2, 3]])].corr()

# By analysing the above correlations we can conclude various things, Insights:
### For eg: Amount invested on wines is positively correlated to income- Those having more income are used to invest more on wines.
### And is negatively correlated to kidhome means those having kids of upto 12yrs are less prone to wines.
### Complain is positively correlated with kidhome and teenhome, hence we can conclude that families having children are having more complains than the others, need to improve the products related to kids and teenagers

# Visualizations

In [ ]:
dupl_df['Marital_Status'].value_counts()  

In [ ]:
#Combining maritial status
dupl_df['Marital_Status'] = dupl_df['Marital_Status'].replace(['Married', 'Together'],'relationship')
dupl_df['Marital_Status'] = dupl_df['Marital_Status'].replace(['Divorced', 'Widow', 'Alone', 'YOLO', 'Absurd'],'Single')

In [ ]:
dupl_df['Marital_Status'].value_counts()  

In [ ]:
# Combining columns
dupl_df['Kids'] = dupl_df['Kidhome'] + dupl_df['Teenhome']
dupl_df['Expenses'] = dupl_df['MntWines'] + dupl_df['MntFruits'] + dupl_df['MntMeatProducts'] + dupl_df['MntFishProducts'] + dupl_df['MntSweetProducts'] + dupl_df['MntGoldProds']
dupl_df['TotalAcceptedCmp'] = dupl_df['AcceptedCmp1'] + dupl_df['AcceptedCmp2'] + dupl_df['AcceptedCmp3'] + dupl_df['AcceptedCmp4'] + dupl_df['AcceptedCmp5'] + dupl_df['Response']
dupl_df['NumTotalPurchases'] = dupl_df['NumWebPurchases'] + dupl_df['NumCatalogPurchases'] + dupl_df['NumStorePurchases'] + dupl_df['NumDealsPurchases']

In [ ]:
#Dropping remaining columns
dupl_df=dupl_df.drop(columns=["AcceptedCmp1" , "AcceptedCmp2", "AcceptedCmp3" , "AcceptedCmp4","AcceptedCmp5", "Response","NumWebVisitsMonth", "NumWebPurchases","NumCatalogPurchases","NumStorePurchases","NumDealsPurchases" , "Kidhome", "Teenhome","MntWines", "MntFruits", "MntMeatProducts", "MntFishProducts", "MntSweetProducts", "MntGoldProds"],axis=1)

In [ ]:
dupl_df

In [ ]:
#AGE
dupl_df["Age"]=2021 - dupl_df["Year_Birth"]

In [ ]:
dupl_df['Education'].value_counts()
dupl_df['Education'] = dupl_df['Education'].replace(['PhD','2n Cycle','Graduation', 'Master'],'PG')  
dupl_df['Education'] = dupl_df['Education'].replace(['Basic'], 'UG')


In [ ]:
dupl_df['Years_Completed'] = 2021- dupl_df['Dt_Customer'].dt.year 

In [ ]:
dupl_df=dupl_df.drop(columns=["ID","Year_Birth","Dt_Customer","Complain"],axis=1)

In [ ]:
dupl_df

In [ ]:
dupl_df.iloc[: , 2 :].describe()

In [ ]:
dupl_df.iloc[: , 2 :].corr()

In [ ]:
import plotly.express as px
px.bar(dupl_df, x='Marital_Status', y='Expenses')

In [ ]:
px.bar(dupl_df, x='Kids', y='Expenses')

In [ ]:
px.histogram(dupl_df,x="Age", facet_row = "Marital_Status")

## Pair plot for numerical columns only

In [ ]:

numerical_cols = dupl_df.select_dtypes(include='number').columns
px.scatter_matrix(dupl_df[numerical_cols])

## 8.

In [ ]:
correlation_matrix = dupl_df[numerical_cols].corr()

# Find top 3 positive correlation pairs
top_3_positive_corr = correlation_matrix.unstack().sort_values(ascending=False)
top_3_positive_corr = top_3_positive_corr[top_3_positive_corr < 1].drop_duplicates().head(3)

# Create scatter plots for the top 3 positive correlation pairs using Plotly
for i in range(0, len(top_3_positive_corr)):
    col1, col2 = top_3_positive_corr.index[i]
    corr_value = top_3_positive_corr[col1, col2]
    fig = px.scatter(dupl_df, x=col1, y=col2, title=f"Scatter plot: {col1} vs {col2} (Correlation: {corr_value:.4f})")
    fig.show()

In [ ]:
correlation_matrix = dupl_df[numerical_cols].corr()

# Find top 3 negative correlation pairs
top_3_positive_corr = correlation_matrix.unstack().sort_values(ascending=True)
top_3_positive_corr = top_3_positive_corr[top_3_positive_corr < 0].drop_duplicates().head(3)

# Create scatter plots for the top 3 negative correlation pairs using Plotly
for i in range(0, len(top_3_positive_corr)):
    col1, col2 = top_3_positive_corr.index[i]
    corr_value = top_3_positive_corr[col1, col2]
    fig = px.scatter(dupl_df, x=col1, y=col2, title=f"Scatter plot: {col1} vs {col2} (Correlation: {corr_value:.4f})")
    fig.show()